Vector of enum cannot be constructed like other vectors

The following code currently fails in JIT with d4a3823,
```
module Foo;

type S = unit {};
type E = enum { A, B, C };

function works1(): vector<uint8> { return vector<uint8>(); }
function works2(): vector<uint8> { return vector<uint8>(); }
function works4(): vector<S> { return vector<S>(); }
function works5(): vector<S> { return []; }

function broken1(): vector<E> { return []; }
function broken2(): vector<E> { return vector<E>(); }
```

```
$ spicyc foo.spicy -j
Foo.cc:122:12: error: no viable conversion from returned value of type 'Vector<Foo::E, (default) std::__1::allocator<__hlt::Foo::E>>' to function return type 'Vector<Foo::E, hilti::rt::vector::Allocator<Foo::E, Foo::E::Undef>>'
    return hilti::rt::Vector<Foo::E>({});
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:183:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'const hilti::rt::Vector<__hlt::Foo::E, hilti::rt::v
ector::Allocator<__hlt::Foo::E, __hlt::Foo::E::Undef> > &' for 1st argument
    Vector(const Vector& other) : V(other) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:184:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'hilti::rt::Vector<__hlt::Foo::E, hilti::rt::vector:
:Allocator<__hlt::Foo::E, __hlt::Foo::E::Undef> > &&' for 1st argument
    Vector(Vector&& other) noexcept : V(std::move(other)) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:186:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'std::initializer_list<E>' for 1st argument
    Vector(std::initializer_list<T> init, const Allocator& alloc = Allocator()) : V(std::move(init), alloc) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:188:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'const rt::List<E> &' for 1st argument
    Vector(const rt::List<T>& l) : std::vector<T>(l.begin(), l.end()) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:189:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'rt::List<E> &&' for 1st argument
    Vector(rt::List<T>&& l) : std::vector<T>(std::move_iterator(l.begin()), std::move_iterator(l.end())) {}
    ^
Foo.cc:127:12: error: no viable conversion from returned value of type 'Vector<Foo::E, (default) std::__1::allocator<__hlt::Foo::E>>' to function return type 'Vector<Foo::E, hilti::rt::vector::Allocator<Foo::E, Foo::E::Undef>>'
    return hilti::rt::Vector<Foo::E>({});
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:183:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'const hilti::rt::Vector<__hlt::Foo::E, hilti::rt::v
ector::Allocator<__hlt::Foo::E, __hlt::Foo::E::Undef> > &' for 1st argument
    Vector(const Vector& other) : V(other) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:184:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'hilti::rt::Vector<__hlt::Foo::E, hilti::rt::vector:
:Allocator<__hlt::Foo::E, __hlt::Foo::E::Undef> > &&' for 1st argument
    Vector(Vector&& other) noexcept : V(std::move(other)) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:186:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'std::initializer_list<E>' for 1st argument
    Vector(std::initializer_list<T> init, const Allocator& alloc = Allocator()) : V(std::move(init), alloc) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:188:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'const rt::List<E> &' for 1st argument
    Vector(const rt::List<T>& l) : std::vector<T>(l.begin(), l.end()) {}
    ^
/usr/local/Cellar/spicy/HEAD-d4a3823/include/hilti/rt/types/vector.h:189:5: note: candidate constructor not viable: no known conversion from 'hilti::rt::Vector<Foo::E>' (aka 'Vector<__hlt::Foo::E>') to 'rt::List<E> &&' for 1st argument
    Vector(rt::List<T>&& l) : std::vector<T>(std::move_iterator(l.begin()), std::move_iterator(l.end())) {}
    ^
2 errors generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit Foo to bitcode
[error] spicyc: JIT compilation failed
```

We probably are missing some handling for enum types when constructing vectors.

--- a/hilti/src/compiler/codegen/coercions.cc
+++ b/hilti/src/compiler/codegen/coercions.cc
@@ -47,8 +47,15 @@ struct Visitor : public hilti::visitor::PreOrder<std::string, Visitor> {
         if ( auto t = dst.tryAs<type::Set>() )
             return fmt("hilti::rt::Set(%s)", expr);
 
-        if ( auto t = dst.tryAs<type::Vector>() )
-            return fmt("hilti::rt::Vector(%s)", expr);
+        if ( auto t = dst.tryAs<type::Vector>() ) {
+            auto x = cg->compile(t->elementType(), codegen::TypeUsage::Storage);
+
+            std::string allocator;
+            if ( auto def = cg->typeDefaultValue(t->elementType()) )
+                allocator = fmt(", hilti::rt::vector::Allocator<%s, %s>", x, *def);
+
+            return fmt("hilti::rt::Vector<%s%s>(%s)", x, allocator, expr);
+        }
 
         logger().internalError(fmt("codegen: unexpected type coercion from lisst to %s", dst.typename_()));
     }

--- a/hilti/src/compiler/codegen/ctors.cc
+++ b/hilti/src/compiler/codegen/ctors.cc
@@ -182,7 +182,13 @@ struct Visitor : hilti::visitor::PreOrder<std::string, Visitor> {
             // Can only be the empty list.
             return "hilti::rt::vector::Empty()";
 
-        return fmt("hilti::rt::Vector<%s>({%s})", cg->compile(n.elementType(), codegen::TypeUsage::Storage),
+        auto x = cg->compile(n.elementType(), codegen::TypeUsage::Storage);
+
+        std::string allocator;
+        if ( auto def = cg->typeDefaultValue(n.elementType()) )
+            allocator = fmt(", hilti::rt::vector::Allocator<%s, %s>", x, *def);
+
+        return fmt("hilti::rt::Vector<%s%s>({%s})", x, allocator,
                    util::join(util::transform(n.value(), [this](auto e) { return fmt("%s", cg->compile(e)); }), ", "));
     }
 

--- a/tests/Baseline/hilti.types.enum.basic/output
+++ b/tests/Baseline/hilti.types.enum.basic/output
@@ -3,3 +3,6 @@ Y::Undef
 X::A1
 Y::B2
 [$x1=X::Undef, $x2=(not set), $y1=(not set)]
+[]
+[]
+[X::A1, X::A2]

--- a/tests/hilti/types/enum/basic.hlt
+++ b/tests/hilti/types/enum/basic.hlt
@@ -38,4 +38,22 @@ assert X::A1;
 assert !X::Undef;
 assert x && X::A1 ;
 
+# Passing back from function
+function vector<X> x1() {
+    return [];
+}
+
+function vector<X> x2() {
+    return vector<X>();
+}
+
+function vector<X> x3() {
+    return [X::A1, X::A2];
+}
+
+
+hilti::print(x1());
+hilti::print(x2());
+hilti::print(x3());
+
 }

